### ISSUE TITLE
Imputed values differ after a kernel is loaded

### ISSUE THREAD

**Issue by KaikeWesleyReis** [AnotherSamWilson/miceforest#51]
Hi @AnotherSamWilson,

The related issue is the fact that for large dataset (my case at least) the imputed values for the same dataset differ before (trained kernel) and after (that same kernel saved and loaded) the imputation. The example bellow show what I'm talking:

![image](https://user-images.githubusercontent.com/32513366/181573436-131bd3b4-f253-49a3-9ac1-1c44a4d100d0.png)

Here 2% was different from one set to another.

I'll reproduce this same idea with Iris dataset to verify.

**Comment by KaikeWesleyReis** [comment]
Hi @AnotherSamWilson , sorry for my delay. I'm not able to reproduce this, but during that time I repeated this 2 or 3 time. It's odd.
I'll update you with a toy data (to check) and then real data.

**Comment by KaikeWesleyReis** [comment]
> Hi @AnotherSamWilson , sorry for my delay. I'm not able to reproduce this, but during that time I repeated this 2 or 3 time. It's odd. I'll update you with a toy data (to check) and then real data.

Hi @AnotherSamWilson , happy new year!

I just came to solve the imputation, and thus I evaluated again and this issue **persists**. Here I'm using in another different type of model (200k samples). The imputation is different if I use a in-memory kernel vs the same kernel, but loaded.

Example of 3 different features:
![ex1](https://user-images.githubusercontent.com/32513366/209011280-a3d824b6-0a56-463c-a0bc-9de4b34efaee.png)

![ex2](https://user-images.githubusercontent.com/32513366/209011335-8c128464-bff3-420d-a385-a0d292179a83.png)

![ex3](https://user-images.githubusercontent.com/32513366/209011377-3a2338e8-df59-49b4-afaa-654f32b9a4e7.png)

The actual code that I'm using:

```
# Define Imputer parameters
mice_datasets = 3
mice_subset = 1.0
mice_iterations = 5

# Define Imputer
mice_imputer = mf.ImputationKernel(data=x_train,
                                   datasets=mice_datasets,
                                   data_subset=mice_subset,
                                   variable_schema=mice_variable_schema,
                                   save_models=2,
                                   random_state=1206,
                                   train_nonmissing=True,
                                   save_all_iterations=True,
                                   save_loggers=True)

# Run the MICE algorithm for X iterations
mice_imputer.mice(iterations=mice_iterations, verbose=True, n_jobs=-1)

# Compile candidate predictions (increase the imputation speed, but now is slower)
# mice_imputer.compile_candidate_preds()

# Impute train / validation / test set
x_train_imp = mice_imputer.impute_new_data(new_data=x_train.copy()).complete_data(dataset=mice_imputer.dataset_count() - 1)

# Save the imputer
mice_imputer.save_kernel('mice_imputer', copy_while_saving=True)

# Load Imputer
mice_imputer_loaded = mf.load_kernel('mice_imputer')

# Impute the inference set
x_train_imp_loaded = mice_imputer_loaded.impute_new_data(new_data=x_train.copy()).complete_data(dataset=mice_imputer_loaded.dataset_count() - 1)
````

My hypothesis: Given the fact that the kernel holds several models, somehow it's not saving the random state of them and thus, all the models have different random conditions. The second hypothesis would be model variance.

I'm using a variable_schema to set which feature should be use to predict the other now, but when I opened this issue, I didn't used.

I'm more than happy to help you to solve this issue. 

**If possible, can you open this issue again?** 

Best regards,

**Comment by KaikeWesleyReis** [comment]
@AnotherSamWilson apparently shall be a problem with not using `random_seed_array`. I'm running some test right now.

Some questions: 

How do I define the size of this array? 

Why it's not guarantee the reproducible imputation at record level?

**Comment by KaikeWesleyReis** [comment]
@AnotherSamWilson My suspicious was true: It's necessary to set `random_seed_array` and `random_state` in `impute_new_data()`.

I just set the same `random_state` value as `random_seed_array`:

```
np.array([1206]*len(x_train), dtype='int32')
```

The predictions were equivalent for loaded and in-memory kernel. Even in-memory kernel had different predictions if I not set the `random_seed_array` and `random_state`.

But, I still have a question: Why this happens? It's a model predicting a value, and I had set the `random_state` at the kernel level, why it's necessary a `random_seed_array`?

**TL;DR = The problem is not the loaded kernel, but the lack of `random_seed_array` and `random_state` in the function `impute_new_data`**

**Comment by AnotherSamWilson** [comment]
Hmmm I would still think that the `random_state` in the kernel should be the same before and after loading. Simply saving and loading a kernel should not produce different imputation values.

`random_seed_array` will definitely ensure the same values are imputed, but the process is somewhat slower so the `random_seed_array` is not used unless specified directly.

I'll need to do some experimentation with a `np.random.RandomState` to see if serialization affects the random state somehow. Without data to reproduce, this will be a tough nut to crack. There are tests that test for this, but apparently it can still happen.

**Comment by KaikeWesleyReis** [comment]
> Hmmm I would still think that the `random_state` in the kernel should be the same before and after loading. Simply saving and loading a kernel should not produce different imputation values.
> 
> `random_seed_array` will definitely ensure the same values are imputed, but the process is somewhat slower so the `random_seed_array` is not used unless specified directly.
> 
> I'll need to do some experimentation with a `np.random.RandomState` to see if serialization affects the random state somehow. Without data to reproduce, this will be a tough nut to crack. There are tests that test for this, but apparently it can still happen.

The problem is associated with using or not the random_seed_array. Doesn't matter if the imputer was loaded or not. That's what I found. I made the following test: given the imputer trained in-memory and the same one saved and loaded I checked:
- Imputation without `random_seed_array` with in-memory
- Imputation with `random_seed_array` with in-memory
- Imputation with `random_seed_array` with loaded

The last two holds the same results. From my perspective, the necessity `random_seed_array` just converted `random_state` useless somehow.
